### PR TITLE
Css Override

### LINKED ISSUE #2157: Generate CSS override

### PR DESCRIPTION
**Pull request by danloa**
Resolves #2157 

- Logos can be uploaded and changed.
- Css are regenerated in runtime.
- Main fonts can be configured.
- To compile the styles the docker image *jbergknoff/sass* is used.
- Notifications were added to notify when css files are recompiled.

image

### CODE REVIEW

File: database/migrations/2019_07_23_185536_create_settings_table.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('settings', function (Blueprint $table) {
+            $table->increments('id');
+            $table->string('key')->index();
+            $table->json('config')->nullable();
```
Comment (velkymx) [inline database/migrations/2019_07_23_185536_create_settings_table.php]: Change to a text field

File: database/migrations/2019_07_23_185536_create_settings_table.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('settings', function (Blueprint $table) {
+            $table->increments('id');
+            $table->string('key')->index();
```
Comment (velkymx) [inline database/migrations/2019_07_23_185536_create_settings_table.php]: Make unique ->unique();

File: database/factories/SettingFactory.php [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+use Faker\Generator as Faker;
+/**
+ * Model factory for a settings.
+ */
+$factory->define(ProcessMaker\Models\Setting::class, function (Faker $faker) {
+    return [
+        'key' => $faker->sentence(1),
```
Comment (velkymx) [inline database/factories/SettingFactory.php]: Add value

File: ProcessMaker/Http/Controllers/Api/CssOverride.php [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+     *
+     * @return array
+     */
+    private function formatConfig(Request $request)
+    {
+        return [
+            'logo' => $request->input('fileLogoName', ''),
+            'icon' => $request->input('fileIconName', ''),
+            'variables' => $request->input('variables', ''),
+            'sansSerifFont' => $request->input('sansSerifFont', $this->sansSerifFontDefault()),
```
Comment (velkymx) [inline ProcessMaker/Http/Controllers/Api/CssOverride.php]: Only one font please

File: resources/views/admin/cssOverride/edit.blade.php [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    <div class="container" id="editCss" v-cloak>
+        <div class="row" role="document">
+            <div class="col">
+                <div class="card card-body">
+                    <div class="form-group">
+                        {!! Form::label('fileLogo', __('Custom Logo')) !!}
+                        <small class="d-block">{{ __('We recommended a transparent PNG at :size pixels.', ['size' => '150x40']) }}</small>
+                        <div class="input-group">
+                            <input type="text" name="fileLogo" class="form-control" v-model="fileLogo.selectedFile"
+                                   placeholder="{{__('Choose logo image')}}">
```
Comment (velkymx) [inline resources/views/admin/cssOverride/edit.blade.php]: Custom ICON

**Comment by PMQABen** [comment]
Things are failing when I run `npm ci` as part of building this branch:
```
npm ERR! Cannot read property 'length' of undefined

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/bfields/.npm/_logs/2019-07-29T21_22_44_431Z-debug.log
```

The debug file:
```
17178 silly buildTree linking accounting@0.4.1
17179 info lifecycle accounting@0.4.1~preinstall: accounting@0.4.1
17180 info lifecycle accounting@0.4.1~install: accounting@0.4.1
17181 info lifecycle accounting@0.4.1~postinstall: accounting@0.4.1
17182 info lifecycle undefined@undefined~install: undefined@undefined
17183 info lifecycle undefined@undefined~postinstall: undefined@undefined
17184 verbose stack TypeError: Cannot read property 'length' of undefined
17184 verbose stack     at _incorrectWorkingDirectory (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:100:60)
17184 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:72:44
17184 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:203:12
17184 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/graceful-fs/polyfills.js:285:20
17184 verbose stack     at FSReqCallback.oncomplete (fs.js:161:5)
17185 verbose cwd /Users/bfields/dev/bpm
17186 verbose Darwin 17.7.0
17187 verbose argv "/usr/local/Cellar/node/11.6.0/bin/node" "/usr/local/bin/npm" "ci"
17188 verbose node v11.6.0
17189 verbose npm  v6.8.0
17190 error Cannot read property 'length' of undefined
17191 verbose exit [ 1, true ]
```

**Comment by Aaron-Cushman** [comment]
--Why is request button linked to success? 
--If you want the side bar and top navigation bar to be the same color, you'll never be able to see the icons on the side bar, or the active link in the navigation bar. 

Customize_UI_-_ProcessMaker

--Custom logo doesn't rescale

Customize_UI_-_ProcessMaker

**Comment by danloa** [comment]
- We added a new dependency to package.json, so it is necessary to use npm install.
- The error message "Cannot read property 'private' of undefined" seems to be caused because Laravel Echo is not running (execute laravel-echo-server start). If this is the case, this is the reason for the message of successful CSS recompile doesn't arrive. To validate this, can you create a process, enable notifications for it and verify if notifications arrive (in the upper right bell)?

**Comment by PMQABen** [comment]
Laravel Echo is running, and process notifications are running fine, but I do not get any confirmation banners for the UI update.

**Comment by danloa** [comment]
@PMQABen  We had some issues when the machine had a firewall active. Can you stop any firewall  (iptables, ufw, etc.)? Additionally, can you restart echo server and queues?

**Comment by velkymx** [comment]
@danloa Getting NPM errors 

npm ERR! Cannot read property 'length' of undefined

```
17183 verbose stack     at _incorrectWorkingDirectory (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:100:60)
17183 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:72:44
17183 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:203:12
17183 verbose stack     at callback (/usr/local/lib/node_modules/npm/node_modules/graceful-fs/polyfills.js:289:20)
17183 verbose stack     at FSReqWrap.oncomplete (fs.js:150:5)
17184 verbose cwd /Users/alanbollinger/dev/bpm
17185 verbose Darwin 18.7.0
17186 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "ci"
17187 verbose node v10.0.0
17188 verbose npm  v6.10.1
17189 error Cannot read property 'length' of undefined
17190 verbose exit [ 1, true ]
```

**Comment by PMQABen** [comment]
> @danloa Getting NPM errors
Run `npm install` instead of `npm ci`

**Comment by PMQABen** [comment]
Problem with color picker/transparency sliders
[rgb.mp4.zip](https://github.com/ProcessMaker/processmaker/files/3449051/rgb.mp4.zip)

**Comment by PMQABen** [comment]
The "style recompile" banner issue was due to socket.io not loading. Fixed it and the problem went away.

**Pull request merged**